### PR TITLE
ARROW-3791: [C++ / Python] Add boolean type inference to the CSV parser

### PR DESCRIPTION
**Pull request by pitrou**
The set of recognized values can be customized using arrow::csv::ConvertOptions.

### CODE REVIEW

File: cpp/src/arrow/csv/converter.cc [hunk 1]
```diff
@@ -183,6 +187,57 @@ Status FixedSizeBinaryConverter::Convert(const BlockParser& parser, int32_t col_
...
+  Trie true_trie_;
+  Trie false_trie_;
+};
+
+Status BooleanConverter::Convert(const BlockParser& parser, int32_t col_index,
+                                 std::shared_ptr<Array>* out) {
+  BooleanBuilder builder(type_, pool_);
+
+  auto visit = [&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
+    // XXX should quoted values be allowed at all?
```
Comment (bkietz) [inline cpp/src/arrow/csv/converter.cc]: Generators could unconditionally emit quoted CSV and still be legal; I think quoting should be transparent to conversion

File: cpp/src/arrow/csv/options.cc [hunk 2]
```diff
@@ -24,10 +24,12 @@ ParseOptions ParseOptions::Defaults() { return ParseOptions(); }
 
 ConvertOptions ConvertOptions::Defaults() {
   auto options = ConvertOptions();
-  // The default list of possible null spellings is taken from Pandas' read_csv().
+  // Same default null / true / false spellings as in Pandas.
   options.null_values = {"",     "#N/A", "#N/A N/A", "#NA",     "-1.#IND", "-1.#QNAN",
                          "-NaN", "-nan", "1.#IND",   "1.#QNAN", "N/A",     "NA",
                          "NULL", "NaN",  "n/a",      "nan",     "null"};
+  options.true_values = {"1", "True", "TRUE", "true"};
```
Comment (bkietz) [inline cpp/src/arrow/csv/options.cc]: Would it make sense to extend `trie` (or maybe just BooleanConverter) to be case insensitive?

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/csv/options.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/csv/options.cc]: In this case, no. Pandas only accepts those exact values. Also, case-insensitive utf-8 is not exactly easy.

File: cpp/src/arrow/csv/converter.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/csv/converter.cc]: Right. It's an open question... If someone complains, we may add a setting :-)

**Review by wesm — APPROVED** [review]
+1 LGTM. Thanks @pitrou!

**Pull request closed without merging**